src/java.base/share/classes/java/lang/invoke/InvokerBytecodeGenerator.java

```diff
@@ -318,227 +320,84 @@ private Object classDataValues() {
-        mv.visitMaxs(2, 1);
-        mv.visitEnd();
-    }
-
-    /*
-     * Low-level emit helpers.
-     */
-    private void emitConst(Object con) {
-        if (con == null) {
-            mv.visitInsn(Opcodes.ACONST_NULL);
-            return;
-        }
-        if (con instanceof Integer) {
-            emitIconstInsn((int) con);
-            return;
-        }
-        if (con instanceof Byte) {
-            emitIconstInsn((byte)con);
-            return;
-        }
-        if (con instanceof Short) {
-            emitIconstInsn((short)con);
-            return;
-        }
-        if (con instanceof Character) {
-            emitIconstInsn((char)con);
-            return;
-        }
-        if (con instanceof Long) {
-            long x = (long) con;
-            short sx = (short)x;
-            if (x == sx) {
-                if (sx >= 0 && sx <= 1) {
-                    mv.visitInsn(Opcodes.LCONST_0 + (int) sx);
+            clb.withField(p.name, p.desc, ACC_STATIC|ACC_FINAL);
+        }
+
+        clb.withMethodBody("<clinit>", MethodTypeDesc.of(CD_void), ACC_STATIC, new Consumer<CodeBuilder>() {
+            @Override
+            public void accept(CodeBuilder cob) {
+                cob.constantInstruction(classDesc)
+                   .invokestatic(CD_MethodHandles, "classData", MethodTypeDesc.of(CD_Object, CD_Class));
+                if (classData.size() == 1) {
+                    ClassData p = classData.get(0);
+                    cob.checkcast(p.desc)
+                       .putstatic(classDesc, p.name, p.desc);
                 } else {
-                    emitIconstInsn((int) x);
-                    mv.visitInsn(Opcodes.I2L);
-                }
-                return;
-            }
-        }
-        if (con instanceof Float) {
-            float x = (float) con;
-            short sx = (short)x;
-            if (x == sx) {
-                if (sx >= 0 && sx <= 2) {
-                    mv.visitInsn(Opcodes.FCONST_0 + (int) sx);
-                } else {
-                    emitIconstInsn((int) x);
-                    mv.visitInsn(Opcodes.I2F);
-                }
-                return;
-            }
-        }
-        if (con instanceof Double) {
-            double x = (double) con;
-            short sx = (short)x;
-            if (x == sx) {
-                if (sx >= 0 && sx <= 1) {
-                    mv.visitInsn(Opcodes.DCONST_0 + (int) sx);
-                } else {
-                    emitIconstInsn((int) x);
-                    mv.visitInsn(Opcodes.I2D);
+                    cob.checkcast(CD_List)
+                       .astore(0);
+                    int index = 0;
+                    for (ClassData p : classData) {
+                        // initialize the static field
+                        cob.aload(0)
+                           .constantInstruction(index++)
+                           .invokeinterface(CD_List, "get", MethodTypeDesc.of(CD_Object, CD_int))
```

Comment:
MTD should be stored as a constant instead of created every time the loop is executed.